### PR TITLE
Add CPU affinity support to UCP worker

### PR DESCRIPTION
**Pull request by guy101**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2617/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/15/ for details.

**Comment by guy101** [comment]
added missing file ucs/type/worker_param.h

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2619/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/17/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/20/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1970/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1972/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/21/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2621/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2622/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1975/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1974/ for details (Mellanox internal link).

File: src/tools/info/proto_info.c [hunk 1]
```diff
@@ -42,8 +42,8 @@ void print_ucp_info(int print_opts, ucs_config_print_flags_t print_flags,
     if (!(print_opts & (PRINT_UCP_WORKER|PRINT_UCP_EP))) {
         goto out_cleanup_context;
     }
-
-    status = ucp_worker_create(context, UCS_THREAD_MODE_MULTI, &worker);
+    ucs_worker_param_t worker_params;
```
Comment (yosefe) [inline src/tools/info/proto_info.c]: worker_params are uninitialized

File: src/tools/info/tl_info.c [hunk 2]
```diff
@@ -208,7 +208,8 @@ static ucs_status_t print_tl_info(uct_md_h md, const char *tl_name,
     }
 
     /* coverity[alloc_arg] */
-    status = uct_worker_create(&async, UCS_THREAD_MODE_MULTI, &worker);
+    ucs_worker_param_t worker_params;
```
Comment (yosefe) [inline src/tools/info/tl_info.c]: worker_params are uninitialized

File: src/ucs/type/worker_param.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 20016.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_TYPE_CPU_PARAM_H
```
Comment (yosefe) [inline src/ucs/type/worker_param.h]: worker_param should be declared in ucp and not in ucs

File: src/ucs/type/worker_param.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#define UCS_TYPE_CPU_PARAM_H
+
+#include <sched.h>
+
+/**
+ * Specifies CPU parameters.
+ */
+typedef struct ucs_worker_param {
+    cpu_set_t cpu_cores;
```
Comment (yosefe) [inline src/ucs/type/worker_param.h]: missing doxygen

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -503,6 +504,7 @@ void uct_release_tl_resource_list(uct_tl_resource_desc_t *resources);
  * @param [out] worker_p      Filled with a pointer to the worker object.
  */
 ucs_status_t uct_worker_create(ucs_async_context_t *async,
+                               ucs_worker_param_t *param,
```
Comment (yosefe) [inline src/uct/api/uct.h]: uct_worker should not have this parameter

File: src/uct/ib/base/ib_iface.c [hunk 6]
```diff
@@ -355,6 +356,17 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
     return UCS_OK;
 }
 
+static int find_lcs_cpu(cpu_set_t *cpu_cores)
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: code style

File: src/uct/ib/base/ib_iface.c [hunk 7]
```diff
@@ -355,6 +356,17 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
...
+{
+    int i = 0;
+    for(i = 0 ;i < CPU_COUNT(cpu_cores);++i) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: - double initialization of i
- code style

File: src/uct/ib/base/ib_iface.c [hunk 8]
```diff
@@ -355,6 +356,17 @@ static ucs_status_t uct_ib_iface_init_lmc(uct_ib_iface_t *iface,
...
+        if(CPU_ISSET(i, cpu_cores)) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: need to use more efficient lookup, e.g bit operations

File: src/uct/ib/base/ib_iface.c [hunk 9]
```diff
@@ -369,6 +381,8 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
     uct_ib_device_t *dev = &ucs_derived_of(md, uct_ib_md_t)->dev;
     ucs_status_t status;
     uint8_t port_num;
+    cpu_set_t cpu_cores    = worker->worker_params.cpu_cores;
+    int used_irq = find_lcs_cpu(&cpu_cores);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: calling this function should not be done in init part, need to move it to right before calling create cq

File: src/uct/ib/base/ib_iface.c [hunk 10]
```diff
@@ -369,6 +381,8 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
     uct_ib_device_t *dev = &ucs_derived_of(md, uct_ib_md_t)->dev;
     ucs_status_t status;
     uint8_t port_num;
+    cpu_set_t cpu_cores    = worker->worker_params.cpu_cores;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: no need for this extra variable

File: test/examples/ucp_hello_world.c [hunk 11]
```diff
@@ -63,6 +64,7 @@ enum ucp_test_mode_t {
 } ucp_test_mode = TEST_MODE_PROBE;
 
 static uint16_t server_port = 13337;
+static long test_string_length = 16;
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: why is modifying the UCP example a part of this PR?

File: src/uct/api/uct.h (hunk 5 above)
Comment (guy101) [inline src/uct/api/uct.h]: please elaborate

File: test/examples/ucp_hello_world.c (hunk 11 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: entered by mistake and will be removed.

File: src/ucs/type/worker_param.h (hunk 3 above)
Comment (guy101) [inline src/ucs/type/worker_param.h]: at first I tried to do it in ucp. I ran into a lot of depenency issues.
hence I asked @shssf for help. He gave me a sample which I used. this is the sample.

Comment (yosefe) [inline src/ucs/type/worker_param.h]: need to not use this structure in UCT at all, so there would not be dependency issues

File: src/uct/api/uct.h (hunk 5 above)
Comment (yosefe) [inline src/uct/api/uct.h]: as we discussed earlier, need to pass cpu number parameter to uct_iface_open instead

**Pull request closed without merging**